### ISSUE TITLE
layout_with_drl  bug: "fixed" is not keeping vertex coordinates fixed.

### ISSUE THREAD

**Issue by nassuphis** [igraph/igraph#916]
the "fixed" parameter  in layout_with_dr()

fixed: Logical vector, it can be used to fix some vertices. All vertices for which it is
TRUE are kept at the coordinates supplied in the seed matrix. It is ignored it
NULL or if use.seed is FALSE

does not seem to work:

``` r
y<-1:10
  ylenl<-length(y)
  g<-graph_from_edgelist(cbind(1:(ylen-1),2:ylen))
  V(g)$x<-as.numeric(seq_along(y))
  V(g)$y<-y*y
  V(g)$name<-paste0("node",seq_along(y))
  seed<-cbind(V(g)$x,V(g)$y)
  rownames(seed)<-V(g)$name
  colnames(seed)<-c("x","y")
  V(g)$fixed<-rep(TRUE,length(V(g)))
  fixed<-V(g)$fixed
  V(g)$weights<-rep(1,length(V(g)))
  weights=V(g)$weights
  set.seed(1)
  #m<-.Call("R_igraph_layout_drl", graph=g, seed=seed, use.seed=TRUE,options=o, weights=weights, fixed=fixed, PACKAGE = "igraph")
  m<-layout_with_drl(g,seed=seed,use.seed=TRUE,fixed=fixed,dim=2)
```

**Comment by nassuphis** [comment]
and "m" comes out to this:

```
> m
           [,1]     [,2]
 [1,] 5.6373711 33.53471
 [2,] 5.4721932 34.21797
 [3,] 5.1933951 34.39255
 [4,] 4.2804580 34.38229
 [5,] 3.5390167 34.99001
 [6,] 2.7601748 35.68501
 [7,] 1.9723243 36.25149
 [8,] 1.2896012 36.96452
 [9,] 0.6291004 37.68113
[10,] 0.1315595 38.08068
```

**Comment by nassuphis** [comment]
while "seed" was this:

```
> seed
        x   y
node1   1   1
node2   2   4
node3   3   9
node4   4  16
node5   5  25
node6   6  36
node7   7  49
node8   8  64
node9   9  81
node10 10 100
```

and we specified "fixed=TRUE" for all vertices and "use.seed=TRUE"

is this the correct behaviour?

**Comment by gaborcsardi** [comment]
Seems like a bug to me.....

**Comment by nassuphis** [comment]
I thought so. had a look, nothing obvious, it seems to be in  igraph.

qgraph has a version with contraints:
https://github.com/SachaEpskamp/qgraph/blob/master/src/layout.c

**Comment by ntamas** [comment]
I have checked the DrL code in the C core of igraph and it seems like the original DrL implementation has never meant to keep the nodes marked in the `fixed` vector as fixed all the way along. Whenever the C code checks whether a node is fixed, it also checks an auxiliary variable named `real_fixed`; setting `real_fixed` to `false` will release the fixed nodes and let them move around freely. Actually, this happens in the DrL code when the total number of iterations performed exceeds a variable named `real_iterations`. If I read the code correctly, `real_iterations` seems to be set to -1, so the nodes never get fixed at all.

I have no idea what the intention of the author of the DrL layout was, so I cannot judge whether this is the expected behaviour (and we just misunderstood what `fixed` meant when we ported DrL to igraph), or this is a bug in the DrL algorithm.

@nassuphis That one is actually the Fruchterman-Reingold layout algorithm; igraph also has it (see `layout_with_fr`) and it also handles constraints for the X and Y coordinates of each vertex.

**Comment by nassuphis** [comment]
ah, you are right, layout_with_fr does allow constraints to be specified by node, i did not read the doc carefully. that will do, thanks.

**Comment by reuning** [comment]
It might be useful to update the docs to say that `fixed ` doesn't do anything . I say this as I just spent several hours trying to figure out what was going on with some code, and it turns out this was the problem.

**Comment by ntamas** [comment]
I'm sorry to hear that; the commit above modifies the documentation accordingly.

**Comment by reuning** [comment]
Thank you very much for this, and for all the other work you put into igraph.

**Comment by vtraag** [comment]
I believe the intention is to only keep the nodes fixed for `real_iterations`. Presumably, this is to make it possible for a node to be fixed initially, but leave it some room in later stages, so that nodes that are fixed tend to remain around their initial positions, but not strictly so. Or at least, that is my best guess. The original software was available from http://www.cs.sandia.gov/~smartin/software.html, but the link is dead unfortunately.

What is the reason not to set `double real_in = 1.0;` in 
https://github.com/szhorvat/igraph/blob/9bb8167ad6d6e5ae3a8359610a85bca57916e4b8/src/drl_graph.cpp#L439
so that effectively nodes remain fixed throughout all stages?

**Comment by iosonofabio** [comment]
@vtraag agreed that seems to be what this is about. The fact that `real_parm == 1` even exists is an indication that people sometimes wanted fixed nodes throughout.

I would be in favour of cleaning up all the dead code in that file and documenting a little better the `real_in` story without touching the API (for now). What do you think?

**Comment by ntamas** [comment]
Usually our policy was that we try to keep upstream code intact so that we can incorporate upstream fixes easily. However, considering that the upstream project is dead, I think there's no point in sticking to this policy any more w.r.t. the DrL codebase. Feel free to tweak or fix whatever you want.

**Comment by vtraag** [comment]
I found a new link: http://www.cs.otago.ac.nz/homepages/smartin/software.php. We can download new source code also. Perhaps we can check out the new source code and plug it in again. Let's do that for 0.9.0 though.

**Comment by ntamas** [comment]
Dev meeting decided that we should remove the `fixed` argument from 0.10 onwards, so it was removed in 925103531 .
